### DISCUSSION TITLE
Unable to use DMA with ST7789 + ESP32-S3

### DISCUSSION THREAD

**Discussion by TheWolf95** [Bodmer/TFT_eSPI#3432] [General]
I only recently come to know about this section of issues/discusson for TFT_eSPI library😂 sorry for my blindess. Getting to the problem (repost of #3414)

IDE: PlatformIO
Library: bodmer/TFT_eSPI@^2.5.43
MCU: ESP32-S3-WROOM-1 N8R8
TFT Driver: ST7789 - SPI (SCK, SDA, RES, DC)

I am trying to use the pushImageDMA() method to achieve faster refreshrate on a ST7789 display, without MISO and CS pin ([link to aliexpress](https://it.aliexpress.com/item/1005005621788608.html)). DMA initialization is ok, but any update to the screen is not sent (no fillScreen, no pushImageDMA, not print, ecc), just the last frame is showed.

Wiring: (it is a custom board from another project, I connected spare pins exposed)

TFT_MOSI 18
TFT_SCLK 17
TFT_DC 2
TFT_RST 1

TFT config:
```
#define ST7789_2_DRIVER

#define TFT_RGB_ORDER TFT_RGB

 #define TFT_WIDTH  240
 #define TFT_HEIGHT  240

#define SPI_FREQUENCY 80*1000000
#define SPI_READ_FREQUENCY  20000000
#define SPI_TOUCH_FREQUENCY  2500000

#define USE_HSPI_PORT
```

Setup before DMA push:
```
if(tft.initDMA()) Serial.println("[DMA] Init ok.");
DMA_BUFFER = (uint16_t *) heap_caps_malloc(256 * sizeof(uint16_t), MALLOC_CAP_DMA);  // These buffers could be 240 only
SCREENBUFFER = (uint16_t *) heap_caps_malloc(256 * sizeof(uint16_t), MALLOC_CAP_DMA);
```

Push in loop function: (this function runs on a parallel thread, priority 2)
```
extern "C" int lcd_write_frame(const uint16_t x, const uint16_t y, const uint16_t width, const uint16_t height) {
...
  for (i = 0; i < height; i++) {
      for (X_ = 0; X_ < width; X_++) { 
          
          SCREENBUFFER[X_] = gbc_16bit[(0x3F & ((uint8_t *)SCREENMEMORY[i])[X_])];

      }

     tft.startWrite();  // I know this should be called to keep the SPI interface reserved through CS pin (not my case), but who knows
     tft.pushImageDMA(0, i, 240, 1, (uint16_t *) SCREENBUFFER, (uint16_t *) DMA_BUFFER);
     tft.endWrite();

  }
...
}
```

I get a refresh times of around 80ms (12 FPS) for 240x240 16bit images, I was aiming on improving them (the code is a fork of GBCCAT, ported to PlatformIO + TFT_eSPI library)

UPDATE:
With some code tinkering I improved the rendering times to 45ms (~22 FPS) witout DMA. The code is similar to the one above tho.
I should add that in the meantime I tried some examples from DMA folder and got no results still. I'm waiting for a devkit with ESP32-WROOM-1 MCU to arrive, to test wether the pin configuration is off

UPDATE 2:
I got hands on a ESP32-S3-DevKit-C1 like board to further test on pin mapping
```
#define TFT_MISO 13
#define TFT_MOSI 11
#define TFT_SCLK 12
#define TFT_CS   10 
#define TFT_DC   38 
#define TFT_RST  2 
#define TFT_BL 9
```
using this setup, with no interface specified (should fall back to VSPI which should be FSPI in S3). I still get the screen freezing to last pushed frame

**Comment by TheWolf95** [comment]
Using the setup as per UPDATE2, I'm able to send a 240x240 buffer via DMA in a simple code
```
#include <TFT_eSPI.h> // Include the graphics library

TFT_eSPI tft = TFT_eSPI();  // Create an instance of the library
uint16_t *dmaBuffer;

void setup() {
  Serial.begin(115200);
  tft.init();
  tft.initDMA();
  tft.setRotation(5);

  tft.setSwapBytes(true);

  // Allocate the DMA buffer
  dmaBuffer = (uint16_t*)heap_caps_malloc(TFT_WIDTH * 240 * sizeof(uint16_t), MALLOC_CAP_DMA | MALLOC_CAP_32BIT);
  if (!dmaBuffer) {
    Serial.println("Failed to allocate DMA buffer!");
    while (1);
  }

  // Fill the DMA buffer with initial data
  for (int i = 0; i < TFT_WIDTH * 240; i++) {
    dmaBuffer[i] = TFT_BLACK; // Initial fill color
  }

  // Perform an initial DMA transfer
  tft.startWrite();
  tft.pushImageDMA(0, 0, TFT_WIDTH, 240, dmaBuffer);
  tft.endWrite();

}

void loop() {
  // Update the DMA buffer with new frame data
  for (int i = 0; i < TFT_WIDTH * 240; i++) {
    dmaBuffer[i] = (i % 3 == 0) ? TFT_RED : ((i % 2 == 0) ? TFT_GREEN : TFT_BLUE); // Example pattern
  }
  Serial.printf("Showing color: 0x%04x\n", dmaBuffer[0]);

  // Perform the DMA transfer
  tft.startWrite();
  tft.pushImageDMA(0, 0, TFT_WIDTH, 240, dmaBuffer);
  tft.endWrite();

  delay(1000);

  tft.fillScreen(TFT_BLACK);

  delay(500);
}
```

MISO (alternatively define `#define TFT_SDA_READ`) and CS must be defined (even randomly) and not -1 and the malloc must be done into a DMA and 32bit accessible part of internal RAM, via the flags above.

I now face another problem in the main code:
```
Free heap: 0, internal: 81328, spiram: 6054475
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.


Core  0 register dump:

PC      : 0x4201b67c  PS      : 0x00060a30  A0      : 0x8201c1bf  A1      : 0x3fcee4e0  
A2      : 0x3fcf7488  A3      : 0x3fc9adc0  A4      : 0x00000000  A5      : 0x00000000  
A6      : 0x60024000  A7      : 0x6002409c  A8      : 0x00000000  A9      : 0x00000000  
A10     : 0x00000000  A11     : 0x00000000  A12     : 0x8037e292  A13     : 0x3fc97400  
A14     : 0x00000000  A15     : 0x00000000  SAR     : 0x0000001d  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000088  LBEG    : 0x400570e8  LEND    : 0x400570f3  LCOUNT  : 0x00000000 


Backtrace: 0x4201b679:0x3fcee4e0 0x4201c1bc:0x3fcee510 0x4200cac3:0x3fcee550 0x4200881c:0x3fcee590 0x42006eb5:0x3fcee5c0


  #0  0x4201b679:0x3fcee4e0 in check_trans_valid at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/driver/spi_master.c:686 (discriminator 2)
  #1  0x4201c1bc:0x3fcee510 in spi_device_queue_trans at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/driver/spi_master.c:803
  #2  0x4200cac3:0x3fcee550 in TFT_eSPI::pushImageDMA(int, int, int, int, unsigned short*, unsigned short*) at .pio/libdeps/esp32-s3-devkitc-1/TFT_eSPI/Processors/TFT_eSPI_ESP32_S3.c:797
  #3  0x4200881c:0x3fcee590 in lcd_write_frame at src/main.cpp:365
  #4  0x42006eb5:0x3fcee5c0 in fb_draw_task at src/GBCCAT/gnuboy.c:256
```

the EXCVADRR value suggests a NULL pointer in a struct, which I suspect being `static spi_transaction_t trans`'s `tx_buffer` pointer in TFT_eSPI class. 
Does anyone have an idea on why the pointer I pass to the `TFT_eSPI::pushImageDMA(int, int, int, int, unsigned short*, unsigned short*)` method is NULL even after I check the malloc beforehand?

UPDATE:
I correct myself, I misread [documentation](https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32/api-guides/fatal-errors.html#loadprohibited-storeprohibited): the error states that the pointer to the struct itself is NULL. So lesser clues than before on what could it be

**Reply by TheWolf95** [reply]
The solution to this problem was within the fact I used `xTask` methods to create tasks to take care of video rendering. The stack size (which still now remains a mistery how to determine its size), was too small, causing the methods executed to not being able to read memory properly (this is what I understood, probably the thruth is similar, but not exact).

TL;DR: are you using `xTaskCreate`? Try increasing the stack size.

I've now to face tearing issues, or better: random pixels are appearing on the first lines of the frame, varying in thickness over time. Curious fact: decreasing the SPI frequency worsen the effect

**Comment by rsiemens77** [comment]
> using this setup, with no interface specified (should fall back to VSPI which should be FSPI in S3). I still get the screen freezing to last pushed frame

> Using the setup as per UPDATE2, I'm able to send a 240x240 buffer via DMA in a simple code

Can you clarify?  So you are not specifying HSPI?  Everything I read says you need to do this, and indeed, unless I do, I get the commonly reported crash.  How did you get it to work without defining HSPI?

Looking at the code, if you do

#define USE_HSPI_PORT

It results in using 

#define SPI_PORT 3     //HSPI is port 3 on ESP32 S2

Is this what you are using?

I can't get my S3 devkit to do DMA, although the non-DMA call works fine.  I used the same pins as you and as listed in the Espressif S3 documentation.  I still get the single frame and then freeze.

Pin Name | GPIO Number (SPI2)

CS0 1 | 10
SCLK | 12
MISO | 13
MOSI | 11
QUADWP | 14
QUADHD | 9

**Reply by TheWolf95** [reply]
Yes, I didn't define HSPI (so it would fall back to VSPI) plus all the other settings and mallocs.
Maybe share your UserSetup and code you are trying to reproduce, so I or someone else can help you further!

**Reply by rsiemens77** [reply]
Thanks, I was following the discussion here https://github.com/Bodmer/TFT_eSPI/discussions/2233 where it was reported working with HSPI.  Let me try your way without HSPI, and then I'll follow up.  Thanks for quick response!

**Reply by rsiemens77** [reply]
I started this thread with my detailed trials and results if you have any insights, I'd appreciate it.

https://github.com/Bodmer/TFT_eSPI/discussions/3473
